### PR TITLE
Persist circuits in batch execute

### LINKED ISSUE #255: Print noisy circuit when a Braket noise model is applied to PL device with `parallel=True`
PennyLane device with Braket backend supports [noise model](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/4560f4e4dc49f621c3698b2d7eeffe6073d09fec/src/braket/pennylane_plugin/braket_device.py#L124). The noise model added to the device will automatically apply to circuit executed by the device.

Normally, one can call `print(device.circuit)` to visualize the circuit after the device has executed it. This allows users to check if the noise is applied as expected. But in the case where the `device` is set to `parallel=True`, this does not work. The underlying issue is that `batch_execute` does not populate the deivce's `circuit` field (because there are multiple circuits being executed on the device).

**Describe the feature you'd like**
If the `device` is initialized with `parallel=True`, the device should cache all of the circuits that have been run on the device so that they can be accessed later. We should also have updated documentation on how to attach a noise model to a device and how to print the resulting circuit(s).

**Describe alternatives you've considered**
The alternative is to use `parallel=False` so that only a single circuit is run and populated in the `circuit` field of the device, and then call `print(device.circuit)`.

### PR DESCRIPTION
**Pull request by WingCode**
*Issue #, if available:* closes #255

*Description of changes:* Add persistence of multiple circuits when batch_execute is called

*Testing done:* 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by math411** [comment]
Hi @WingCode and thanks for raising this! It seems the linter check is failing. Locally, you can run `tox -e linters` to fix this.

**Comment by WingCode** [comment]
@math411 Could you please review?

File: src/braket/pennylane_plugin/braket_device.py [hunk 1]
```diff
@@ -584,6 +586,7 @@ def __init__(
         self._max_parallel = max_parallel
         self._max_connections = max_connections
         self._max_retries = max_retries
+        self.circuits = []
```
Comment (yitchen-tim) [inline src/braket/pennylane_plugin/braket_device.py:589]:
```suggestion
        self._circuits = []
```

File: src/braket/pennylane_plugin/braket_device.py [hunk 2]
```diff
@@ -621,6 +624,7 @@ def batch_execute(self, circuits, **run_kwargs):
                     **run_kwargs,
                 )
             )
+            self.circuits.append(circuit)
```
Comment (yitchen-tim) [inline src/braket/pennylane_plugin/braket_device.py:627]:
```suggestion
        self._circuits = braket_circuits
```

File: src/braket/pennylane_plugin/braket_device.py [hunk 3]
```diff
@@ -139,6 +139,7 @@ def __init__(
         super().__init__(wires, shots=shots or None)
         self._device = device
         self._circuit = None
+        self._circuits = []
```
Comment (yitchen-tim) [inline src/braket/pennylane_plugin/braket_device.py:142]: `self._circuits` here is a private variable. that users should not access. Instead, users access _circuits through properties. So, there need to be a new property of this class called `circuits` defined as a method of the class, similar to  https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/a17e8b99c8089dde79786729ad482af08e3b3431/src/braket/pennylane_plugin/braket_device.py#L174.

Comment (yitchen-tim) [reply src/braket/pennylane_plugin/braket_device.py:142]: And yes, it's a bit repetitive now that we have both `circuits` and `circuit`, but I think it's okay for now. In long term, `circuit` may be deprecated.

File: src/braket/pennylane_plugin/braket_device.py (hunk 2 above)
Comment (yitchen-tim) [reply src/braket/pennylane_plugin/braket_device.py:627]: ^ Note: this suggestion also moves the code out of the for loop

**Review by yitchen-tim — CHANGES_REQUESTED** [review]
Hi @WingCode , thanks for contributing to the issue! This PR solves half of the problem stated in https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/issues/255. Printing noisy circuits when using noise model is still not supported. However, after reviewing this PR, I think printing noisy circuits is not the right goal to pursue.

The right goal is to allow users to retrieve the tasks (and the task ids), from which noisy circuits can be retrieved. If you can extend this PR to include the following feature, we are good to go: "Persist the task when parallel=True", which requires

1. Add an attribute `self._tasks=[]`
2. Add a method for returning tasks
```
    @property
    def tasks(self) -> list[QuantumTask]:
        ...
``` 
, similar to 
https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/a17e8b99c8089dde79786729ad482af08e3b3431/src/braket/pennylane_plugin/braket_device.py#L179
3. Add unit tests for `tasks`

**Review by yitchen-tim — APPROVED** [review]
Hi @WingCode, the PR looks good! I changed the target branch to be a feature branch instead of the main branch because Amazon Braket needs rework the duplication of `circuit` and `circuits` which requires some discussions. I will assign the issue to you and close the issue. This should be good in UnitaryHack regards. Thanks for contributing! 

 I somehow have trouble assigning the issue to you. I am asking my team to resolve it.

**Comment by WingCode** [comment]
Thank you @yitchen-tim for the reviews and all of the timely support!

**Comment by yitchen-tim** [comment]
@WingCode Could you leave a comment on the issue: https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/issues/255? After that, I should be able to assign you.

**Pull request merged**